### PR TITLE
Add `LazyList.mapWithIndex`

### PR DESCRIPTION
**Pull request by jaschdoc**
Resolves #2033

### CODE REVIEW

**Comment by magnus-madsen** [comment]
As a work around, does something like this work:

```
pub def mapWithIndex(f: (a, Int32) -> b & ef, l: LazyList[a]): LazyList[b] & ef =
    let g = match (x, i) -> f(x, i);
    matchEff g {
        case Pure   => mapWithIndexL(f as ((a, Int32)) -> b & Pure,   l)
        case Impure => mapWithIndexE(f as ((a, Int32)) -> b & Impure, l)
    } as & ef

def mapWithIndexL(f: ((a, Int32)) -> b, l: LazyList[a]): LazyList[b] = ???

def mapWithIndexE(f: ((a, Int32)) -> b & Impure, l: LazyList[a]): LazyList[b] & Impure = ???
```

?

**Comment by jaschdoc** [comment]
Hmm... I get this stack trace currently. Any clue what it means?

```
class TestLazyList.Clo$mapWithIndexFusion01$463633 cannot be cast to class Fn1$Obj$Int32 (TestLazyList.Clo$mapWithIndexFusion01$463633 and Fn1$Obj$Int32 are in unnamed module of loader ca.uwaterloo.flix.language.phase.jvm.FlixClassLoader @b2c5e07)
java.lang.ClassCastException: class TestLazyList.Clo$mapWithIndexFusion01$463633 cannot be cast to class Fn1$Obj$Int32 (TestLazyList.Clo$mapWithIndexFusion01$463633 and Fn1$Obj$Int32 are in unnamed module of loader ca.uwaterloo.flix.language.phase.jvm.FlixClassLoader @b2c5e07)
	at LazyList.Def$mapWithIndex$375107.invoke(Unknown Source)
	at TestLazyList.Def$mapWithIndexFusion01.invoke(Unknown Source)
	at TestLazyList.Ns.m_mapWithIndexFusion01(Unknown Source)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at ca.uwaterloo.flix.language.phase.jvm.JvmBackend$.$anonfun$link$1(JvmBackend.scala:276)
	at ca.uwaterloo.flix.language.phase.jvm.JvmBackend$.$anonfun$getCompiledDefs$2(JvmBackend.scala:249)
	at ca.uwaterloo.flix.util.FlixTest.$anonfun$runTests$8(FlixTest.scala:82)
	at org.scalatest.Assertions.withClue(Assertions.scala:1223)
	at org.scalatest.Assertions.withClue$(Assertions.scala:1210)
	at org.scalatest.FunSuite.withClue(FunSuite.scala:1560)
	at ca.uwaterloo.flix.util.FlixTest.$anonfun$runTests$7(FlixTest.scala:80)
	at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:186)
	at org.scalatest.TestSuite.withFixture(TestSuite.scala:196)
	at org.scalatest.TestSuite.withFixture$(TestSuite.scala:195)
	at org.scalatest.FunSuite.withFixture(FunSuite.scala:1560)
	at org.scalatest.FunSuiteLike.invokeWithFixture$1(FunSuiteLike.scala:184)
	at org.scalatest.FunSuiteLike.$anonfun$runTest$1(FunSuiteLike.scala:196)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:286)
	at org.scalatest.FunSuiteLike.runTest(FunSuiteLike.scala:196)
	at org.scalatest.FunSuiteLike.runTest$(FunSuiteLike.scala:178)
	at org.scalatest.FunSuite.runTest(FunSuite.scala:1560)
	at org.scalatest.FunSuiteLike.$anonfun$runTests$1(FunSuiteLike.scala:229)
	at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:393)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:381)
	at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:376)
	at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:458)
	at org.scalatest.FunSuiteLike.runTests(FunSuiteLike.scala:229)
	at org.scalatest.FunSuiteLike.runTests$(FunSuiteLike.scala:228)
	at org.scalatest.FunSuite.runTests(FunSuite.scala:1560)
	at org.scalatest.Suite.run(Suite.scala:1124)
	at org.scalatest.Suite.run$(Suite.scala:1106)
	at org.scalatest.FunSuite.org$scalatest$FunSuiteLike$$super$run(FunSuite.scala:1560)
	at org.scalatest.FunSuiteLike.$anonfun$run$1(FunSuiteLike.scala:233)
	at org.scalatest.SuperEngine.runImpl(Engine.scala:518)
	at org.scalatest.FunSuiteLike.run(FunSuiteLike.scala:233)
	at org.scalatest.FunSuiteLike.run$(FunSuiteLike.scala:232)
	at org.scalatest.FunSuite.run(FunSuite.scala:1560)
	at org.scalatest.Suite.callExecuteOnSuite$1(Suite.scala:1187)
	at org.scalatest.Suite.$anonfun$runNestedSuites$1(Suite.scala:1234)
	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
	at org.scalatest.Suite.runNestedSuites(Suite.scala:1232)
	at org.scalatest.Suite.runNestedSuites$(Suite.scala:1166)
	at org.scalatest.Suites.runNestedSuites(Suites.scala:56)
	at org.scalatest.Suite.run(Suite.scala:1121)
	at org.scalatest.Suite.run$(Suite.scala:1106)
	at org.scalatest.Suites.run(Suites.scala:56)
	at org.scalatest.tools.SuiteRunner.run(SuiteRunner.scala:45)
	at org.scalatest.tools.Runner$.$anonfun$doRunRunRunDaDoRunRun$13(Runner.scala:1349)
	at org.scalatest.tools.Runner$.$anonfun$doRunRunRunDaDoRunRun$13$adapted(Runner.scala:1343)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:1343)
	at org.scalatest.tools.Runner$.$anonfun$runOptionallyWithPassFailReporter$24(Runner.scala:1033)
	at org.scalatest.tools.Runner$.$anonfun$runOptionallyWithPassFailReporter$24$adapted(Runner.scala:1011)
	at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:1509)
	at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:1011)
	at org.scalatest.tools.Runner$.run(Runner.scala:850)
	at org.scalatest.tools.Runner.run(Runner.scala)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.runScalaTest2or3(ScalaTestRunner.java:38)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.main(ScalaTestRunner.java:25)
```

**Comment by magnus-madsen** [comment]
I think it means that a cast is wrong.

**Comment by magnus-madsen** [comment]
If it does not readily work, we can try some other work around.

**Comment by magnus-madsen** [comment]
Like pass a record `{x: a, i: Int32}`

**Comment by jaschdoc** [comment]
I know it changes the semantics slightly, but could this be viable?
```
    pub def mapWithIndex(f: a -> (Int32 -> b & ef), l: LazyList[a]): LazyList[b] & ef =
        matchEff f {
            case Pure   => mapWithIndexL(f as a -> (Int32 -> b & Pure),   l)
            case Impure => mapWithIndexE(f as a -> (Int32 -> b & Impure), l)
        } as & ef
```

**Comment by magnus-madsen** [comment]
No, I don't want to change the signature of mapWithIndex.

File: main/src/library/LazyList.flix [hunk 1]
```diff
@@ -420,7 +414,7 @@ namespace LazyList {
     ///
     /// The function `f` must be pure.
     ///
-    pub def exists(f: a -> Bool & ef, l: LazyList[a]): Bool & ef = match l {
+    pub def exists(f: a -> Bool, l: LazyList[a]): Bool = match l {
```
Comment (magnus-madsen) [inline main/src/library/LazyList.flix:417]: I think these should be allowed to be effect polymorphic.

File: main/src/library/LazyList.flix [hunk 2]
```diff
@@ -434,7 +428,7 @@ namespace LazyList {
     ///
     /// The function `f` must be pure.
     ///
-    pub def forall(f: a -> Bool & ef, l: LazyList[a]): Bool & ef = match l {
+    pub def forall(f: a -> Bool, l: LazyList[a]): Bool = match l {
```
Comment (magnus-madsen) [inline main/src/library/LazyList.flix:431]: I think these should be allowed to be effect polymorphic.

File: main/src/library/LazyList.flix (hunk 1 above)
Comment (jaschdoc) [reply main/src/library/LazyList.flix:417]: Ah yeah I see there were some merge conflicts. They should definitely be effect polymorphic.

**Pull request merged**